File: MangaViewer/app/src/main/java/com/king/mangaviewer/ui/main/fragment/FavouriteFragment.java
package com.king.mangaviewer.ui.main.fragment;


import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.king.mangaviewer.R;
import com.king.mangaviewer.activity.BaseFragment;
import com.king.mangaviewer.adapter.FavouriteMangaItemAdapter;
import com.king.mangaviewer.model.FavouriteMangaMenuItem;

import com.king.mangaviewer.ui.main.MainActivity;
import java.util.Collections;
import java.util.List;


public class FavouriteFragment extends BaseFragment {

    private RecyclerView mRecyclerView;
    private GridLayoutManager gridLayoutManager;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private TextView tv;
    private List<FavouriteMangaMenuItem> dateList = null;

    public FavouriteFragment() {
        this.setHasOptionsMenu(true);
    }

    @Override
    public void onResume() {
        //Toast.makeText(getActivity(),"OnResume",Toast.LENGTH_SHORT);
//        getFavouriteMangaList();
        super.onResume();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_favourite, container, false);
        mRecyclerView = (RecyclerView) rootView.findViewById(R.id.viewPager);
        gridLayoutManager = new GridLayoutManager(getActivity(), getResources().getInteger(R.integer.gridvivew_column_num));
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(gridLayoutManager);

        tv = (TextView) rootView.findViewById(R.id.textView);
        mSwipeRefreshLayout = (SwipeRefreshLayout) rootView.findViewById(R.id.swipeRefreshLayout);
        mSwipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                startAsyncTask();
            }
        });

        startAsyncTask();
        return rootView;
    }

    @Override
    protected Void getContentBackground() {
        getFavouriteMangaList();
        return super.getContentBackground();
    }

    @Override
    protected void updateContent() {
        super.updateContent();
        MainActivity copy = (MainActivity) getActivity();

        FavouriteMangaItemAdapter rcAdapter = new FavouriteMangaItemAdapter(copy, copy.getAppViewModel().Manga, dateList);
        mRecyclerView.setAdapter(rcAdapter);

        tv.setVisibility(View.GONE);
        if (dateList != null && dateList.size() == 0) {
            tv.setText(getString(R.string.favourite_no_favourite_manga));
            tv.setVisibility(View.VISIBLE);
        }

        mSwipeRefreshLayout.setRefreshing(false);
    }

    private void getFavouriteMangaList() {
        MainActivity copy = (MainActivity) getActivity();
        dateList = copy.getAppViewModel().Setting.getFavouriteMangaList();
        Collections.sort(dateList);
        Collections.reverse(dateList);

    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (gridLayoutManager != null) {
            gridLayoutManager.setSpanCount(getResources().getInteger(R.integer.gridvivew_column_num));
        }
    }
}

File: MangaViewer/app/src/main/java/com/king/mangaviewer/ui/main/fragment/FavouriteFragment.kt
package com.king.mangaviewer.ui.main.fragment

import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.util.Pair
import androidx.lifecycle.Observer
import com.king.mangaviewer.R
import com.king.mangaviewer.R.integer
import com.king.mangaviewer.adapter.FavouriteMangaItemAdapter
import com.king.mangaviewer.adapter.MangaMenuItemAdapter.MangaMenuAdapterListener
import com.king.mangaviewer.base.BaseFragment
import com.king.mangaviewer.base.ViewModelFactory
import com.king.mangaviewer.di.annotation.FragmentScopedFactory
import com.king.mangaviewer.model.LoadingState.Idle
import com.king.mangaviewer.model.LoadingState.Loading
import com.king.mangaviewer.model.MangaMenuItem
import com.king.mangaviewer.util.AppNavigator
import com.king.mangaviewer.util.Logger
import com.king.mangaviewer.util.withViewModel
import javax.inject.Inject

class FavouriteFragment : BaseFragment() {

    private lateinit var mRecyclerView: androidx.recyclerview.widget.RecyclerView
    private lateinit var gridLayoutManager: androidx.recyclerview.widget.GridLayoutManager
    private lateinit var mSwipeRefreshLayout: androidx.swiperefreshlayout.widget.SwipeRefreshLayout
    private lateinit var tv: TextView

    lateinit var viewModel: FavouriteFragmentViewModel

    @Inject
    @field:FragmentScopedFactory
    lateinit var fragmentViewModelFactory: ViewModelFactory

    @Inject
    lateinit var appNavigator: AppNavigator

    init {
        this.setHasOptionsMenu(true)
    }

    override fun onResume() {
        super.onResume()
        viewModel.refresh(true)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        menu!!.clear()
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?): View? {

        val rootView = inflater.inflate(R.layout.fragment_favourite, container, false)
        mRecyclerView = rootView.findViewById<View>(
            R.id.viewPager) as androidx.recyclerview.widget.RecyclerView
        gridLayoutManager = androidx.recyclerview.widget.GridLayoutManager(activity,
            resources.getInteger(integer.gridvivew_column_num))
        mRecyclerView.layoutManager = gridLayoutManager
        mRecyclerView.adapter = FavouriteMangaItemAdapter(object : MangaMenuAdapterListener {
            override fun onItemClicked(view: View, item: MangaMenuItem) {
                viewModel.selectMangaMenu(item)
                appNavigator.navigateToChapter(Pair(view, "cover"))
            }
        })

        tv = rootView.findViewById<View>(R.id.textView) as TextView
        mSwipeRefreshLayout = rootView.findViewById<View>(
            R.id.swipeRefreshLayout) as androidx.swiperefreshlayout.widget.SwipeRefreshLayout
        mSwipeRefreshLayout.setOnRefreshListener {
            viewModel.refresh(false)
        }
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViewModel()
    }

    private fun initViewModel() {
        withViewModel<FavouriteFragmentViewModel>(fragmentViewModelFactory) {
            viewModel = this
            this.loadingState.observe(this@FavouriteFragment, Observer {
                when (it) {
                    is Loading -> {
                        mSwipeRefreshLayout.isRefreshing = true
                    }
                    is Idle -> {
                        mSwipeRefreshLayout.isRefreshing = false
                    }
                }
            })

            this.mangaList.observe(this@FavouriteFragment, Observer {
                Logger.d(TAG, "manga list observe, size: ${it!!.size}")
                if (it!!.isEmpty()) {
                    tv.visibility = View.VISIBLE
                } else {
                    tv.visibility = View.GONE
                }
                (mRecyclerView.adapter as? FavouriteMangaItemAdapter)?.submitList(it)

            })

            this.attachToView()
        }

    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        gridLayoutManager.spanCount = resources.getInteger(R.integer.gridvivew_column_num)
    }

    companion object {
        const val TAG = "FavouriteFragment"
    }
}
